File: lib/models/campaign.dart
/// Sovrinli sezonli reyting (aksiya) modellari.
///
/// Doimiy reytingdan (BarberShop.rating) FARQI:
///  - yulduz soni ahamiyatsiz: har bir odam 1 ta OVOZ beradi
///  - bir foydalanuvchi butun aksiya davomida FAQAT BITTA provayderni
///    tanlaydi
///  - aksiya belgilangan kundan boshlanadi va tugaydi
///
/// Backend mos endpointlari:
///   GET  /api/v1/campaigns/active
///   GET  /api/v1/campaigns/{id}/leaderboard
///   GET  /api/v1/campaigns/{id}/my-vote
///   POST /api/v1/campaigns/{id}/vote
library;

enum CampaignStatus { upcoming, running, finished, disabled }

CampaignStatus campaignStatusFrom(String? raw) {
  switch (raw) {
    case 'running':
      return CampaignStatus.running;
    case 'upcoming':
      return CampaignStatus.upcoming;
    case 'finished':
      return CampaignStatus.finished;
    default:
      return CampaignStatus.disabled;
  }
}

class Campaign {
  final int id;
  final String title;
  final String? description;
  final int? categoryId;
  final DateTime startsAt;
  final DateTime endsAt;
  final String? prize;
  final CampaignStatus status;

  /// True bo'lsa faqat o'sha provayderda yakunlangan buyurtmasi bor
  /// foydalanuvchi ovoz bera oladi (soxta ovozga qarshi).
  final bool requireCompletedOrder;

  const Campaign({
    required this.id,
    required this.title,
    required this.startsAt,
    required this.endsAt,
    required this.status,
    this.description,
    this.categoryId,
    this.prize,
    this.requireCompletedOrder = true,
  });

  bool get isRunning => status == CampaignStatus.running;

  /// Tugashiga qancha qoldi (tugagan bo'lsa Duration.zero).
  Duration get remaining {
    final left = endsAt.difference(DateTime.now());
    return left.isNegative ? Duration.zero : left;
  }

  String get remainingLabel {
    if (status == CampaignStatus.finished) return 'Yakunlandi';
    if (status == CampaignStatus.upcoming) return 'Tez orada';
    if (status == CampaignStatus.disabled) return "To'xtatilgan";
    final d = remaining;
    if (d.inDays > 0) return '${d.inDays} kun qoldi';
    if (d.inHours > 0) return '${d.inHours} soat qoldi';
    return '${d.inMinutes} daqiqa qoldi';
  }

  factory Campaign.fromJson(Map<String, dynamic> json) => Campaign(
        id: json['id'] as int,
        title: json['title'] as String,
        description: json['description'] as String?,
        categoryId: json['category_id'] as int?,
        startsAt: DateTime.parse(json['starts_at'] as String),
        endsAt: DateTime.parse(json['ends_at'] as String),
        prize: json['prize'] as String?,
        status: campaignStatusFrom(json['status'] as String?),
        requireCompletedOrder:
            json['require_completed_order'] as bool? ?? true,
      );

}

/// Reyting qatori: provayder + shu aksiyadagi ovozlar soni.
class CampaignRanking {
  final int providerId;
  final String name;
  final String address;
  final int votes;
  final int position;

  const CampaignRanking({
    required this.providerId,
    required this.name,
    required this.address,
    required this.votes,
    required this.position,
  });

  factory CampaignRanking.fromJson(Map<String, dynamic> json) =>
      CampaignRanking(
        providerId: json['id'] as int,
        name: json['name'] as String,
        address: json['address'] as String,
        votes: json['votes'] as int,
        position: json['position'] as int,
      );

}

File: backend/app/models/campaign.py
"""Sezonli reyting (aksiya) modellari.

MAQSAD: doimiy o'rtacha reyting (Provider.rating) yonida VAQT BILAN
CHEGARALANGAN musobaqa. Masalan "Eng yaxshi sartarosh — Sentabr 2026":
belgilangan sanadan boshlanadi, foydalanuvchilar ovoz beradi, g'olibga
sovrin.

DOIMIY REYTINGDAN FARQI:
- Yulduz soni AHAMIYATSIZ: 1 yulduz ham, 5 yulduz ham 1 ta OVOZ.
  Musobaqada "kim ko'proq mijoz yig'gan" muhim, "kim chiroyliroq
  baholangan" emas.
- Bir foydalanuvchi bitta aksiyada FAQAT BIR MARTA ovoz beradi. Bu DB
  darajasida UNIQUE cheklov bilan ta'minlanadi (ilova mantig'i emas) —
  bir vaqtda ikkita so'rov kelsa ham ikkinchisi DB tomonidan rad etiladi.
- Natija aksiya tugagach ham saqlanib qoladi (arxiv).
"""

from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Campaign(Base):
    """Sovrinli reyting aksiyasi (bitta sezon)."""

    __tablename__ = "campaigns"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    title: Mapped[str] = mapped_column(String(200))
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Qaysi yo'nalish bo'yicha musobaqa (sartaroshlar, futbol maydonlari...).
    # NULL = barcha kategoriyalar qatnashadi.
    category_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("categories.id", ondelete="SET NULL"),
        nullable=True, index=True,
    )

    # Aksiya oynasi. Ovoz faqat shu oraliqda qabul qilinadi.
    starts_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)
    ends_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)

    # Sovrin tavsifi (erkin matn: "1-o'rin: 5 000 000 so'm")
    prize: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Admin aksiyani vaqtincha to'xtatib turishi mumkin (sanani
    # o'zgartirmasdan). Faol bo'lishi uchun is_active=True VA hozirgi
    # vaqt [starts_at, ends_at] oralig'ida bo'lishi kerak.
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)

    # SOXTA OVOZGA QARSHI: True bo'lsa faqat o'sha provayderda YAKUNLANGAN
    # buyurtmasi bor foydalanuvchi ovoz bera oladi.
    #
    # Sovrin pul bo'lgani uchun bu jiddiy: aks holda soxta akkauntlar
    # bilan ovoz yig'ish mumkin. Loyihaning sharh tizimi ham xuddi shu
    # qoidani qo'llaydi (provider_service.add_review) va provider_fraud_stats
    # modeli bor — demak soxta baho bu yerda tanilgan muammo.
    #
    # False qilish mumkin: masalan ochiq "xalq ovozi" tanlovi uchun.
    require_completed_order: Mapped[bool] = mapped_column(
        Boolean, default=True, server_default="true"
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    category = relationship("Category")
    votes = relationship(
        "CampaignVote", back_populates="campaign",
        cascade="all, delete-orphan", lazy="selectin",
    )

    def status(self, now: datetime | None = None) -> str:
        """upcoming | running | finished | disabled"""
        if not self.is_active:
            return "disabled"
        now = now or datetime.now(self.starts_at.tzinfo)
        if now < self.starts_at:
            return "upcoming"
        if now > self.ends_at:
            return "finished"
        return "running"

    def to_dict(self, now: datetime | None = None) -> dict:
        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "category_id": self.category_id,
            "starts_at": self.starts_at.isoformat() if self.starts_at else None,
            "ends_at": self.ends_at.isoformat() if self.ends_at else None,
            "prize": self.prize,
            "is_active": self.is_active,
            "require_completed_order": self.require_completed_order,
            "status": self.status(now),
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }


class CampaignVote(Base):
    """Bitta ovoz: foydalanuvchi -> provayder, ma'lum aksiya doirasida."""

    __tablename__ = "campaign_votes"
    __table_args__ = (
        # ENG MUHIM CHEKLOV: bitta aksiyada bitta foydalanuvchi = bitta ovoz.
        # Ilova mantig'iga tayanmaymiz -- bir vaqtda kelgan ikkita so'rovni
        # faqat DB to'g'ri hal qiladi (race condition).
        UniqueConstraint("campaign_id", "user_id", name="uq_campaign_vote_once"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    campaign_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("campaigns.id", ondelete="CASCADE"), index=True
    )
    user_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    provider_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("providers.id", ondelete="CASCADE"), index=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    campaign = relationship("Campaign", back_populates="votes")
    provider = relationship("Provider")
    user = relationship("User")

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "campaign_id": self.campaign_id,
            "user_id": self.user_id,
            "provider_id": self.provider_id,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }
